bd-session-replay/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+  fn take_screenshot(&self);
+}
+
+//
+// Recorder
+//
+
+pub struct Recorder {
+  target: Box<dyn Target + Send + Sync>,
+
+  is_periodic_reporting_enabled: bool,
+  reporting_interval_rate: Duration,
+  reporting_interval: Option<Interval>,
+
+  is_screenshotting_enabled: bool,
+  take_screenshot_rx: Receiver<()>,
+
+  is_periodic_reporting_enabled_flag: BoolWatch<session_replay::PeriodicWireframesEnabledFlag>,
+  reporting_interval_flag: DurationWatch<session_replay::ReportingIntervalFlag>,
+  is_screenshotting_enabled_flag: BoolWatch<session_replay::ScreenshotsEnabledFlag>,
+}
+
+impl Recorder {
+  pub fn new(
+    target: Box<dyn Target + Send + Sync>,
+    runtime_loader: &Arc<ConfigLoader>,
+  ) -> (Self, Sender<()>) {
+    let (take_screenshot_tx, take_screenshot_rx) = channel(2);
+
+    let mut is_periodic_reporting_enabled =
+      session_replay::PeriodicWireframesEnabledFlag::register(runtime_loader).unwrap();
+    let reporting_interval_rate = session_replay::ReportingIntervalFlag::register(runtime_loader)
+      .unwrap()
+      .read_mark_update();
+
+    let mut is_screenshotting_enabled_flag =
+      session_replay::ScreenshotsEnabledFlag::register(runtime_loader).unwrap();
+
+    (
+      Self {
+        target,
+        is_periodic_reporting_enabled: is_periodic_reporting_enabled.read_mark_update(),
+        reporting_interval_rate: reporting_interval_rate.unsigned_abs(),
+        reporting_interval: None,
+        is_screenshotting_enabled: is_screenshotting_enabled_flag.read_mark_update(),
+        take_screenshot_rx,
+        is_periodic_reporting_enabled_flag:
+          session_replay::PeriodicWireframesEnabledFlag::register(runtime_loader).unwrap(),
+        reporting_interval_flag: session_replay::ReportingIntervalFlag::register(runtime_loader)
+          .unwrap(),
+        is_screenshotting_enabled_flag,
+      },
+      take_screenshot_tx,
+    )
+  }
+
+  fn create_interval(interval: Duration, fire_immediately: bool) -> tokio::time::Interval {
+    let mut interval = if fire_immediately {
+      tokio::time::interval(interval)
+    } else {
+      tokio::time::interval_at(Instant::now() + interval, interval)
+    };
+
+    log::debug!("session replay interval is {:?}", interval.period());
+
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+    interval
+  }
+
+  pub async fn run(&mut self) {
+    let shutdown_trigger = ComponentShutdownTrigger::default();
+    self
+      .run_with_shutdown(shutdown_trigger.make_shutdown())
+      .await;
+  }
+
+  pub async fn run_with_shutdown(&mut self, mut shutdown: ComponentShutdown) {
+    if self.reporting_interval.is_none() {
+      self.reporting_interval = Some(Self::create_interval(self.reporting_interval_rate, true));
+    }
+
+    let local_shutdown = shutdown.cancelled();
+    tokio::pin!(local_shutdown);
+
+    loop {
+      tokio::select! {
+        () = async {
+          if let Some(reporting_interval) = &mut self.reporting_interval {
```

Comment:
Just unwrap in here, the outer guard makes it impossible to get in here without it being is_some()

Reply:
ping on this